File: smart-mall-spring/src/main/java/com/example/smart_mall_spring/Services/Logistics/GhtkService.java
package com.example.smart_mall_spring.Services.Logistics;

import com.example.smart_mall_spring.Entities.Address;
import com.example.smart_mall_spring.Entities.Logistics.ShipmentLog;
import com.example.smart_mall_spring.Entities.Logistics.ShipmentOrder;
import com.example.smart_mall_spring.Entities.Orders.Order;
import com.example.smart_mall_spring.Entities.Orders.OrderItem;
import com.example.smart_mall_spring.Entities.Products.ProductVariant;
import com.example.smart_mall_spring.Entities.Users.UserAddress;
import com.example.smart_mall_spring.Enum.ShipmentStatus;
import com.example.smart_mall_spring.Repositories.Logistics.ShipmentLogRepository;
import com.example.smart_mall_spring.Repositories.Logistics.ShipmentOrderRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.*;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestTemplate;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

@Service
@RequiredArgsConstructor
public class GhtkService {

    private final RestTemplate restTemplate = new RestTemplate();
    private final ShipmentOrderRepository shipmentOrderRepository;
    private final ShipmentLogRepository shipmentLogRepository;

    @Value("${ghtk.token}")
    private String TOKEN;

    @Value("${ghtk.partner-code}")
    private String PARTNER_CODE;

    private static final String BASE_URL = "https://services.giaohangtietkiem.vn/services/";

    private HttpHeaders defaultHeaders() {
        HttpHeaders headers = new HttpHeaders();
        headers.set("Token", TOKEN);
        headers.set("X-Client-Source", PARTNER_CODE);
        headers.setContentType(MediaType.APPLICATION_JSON);
        return headers;
    }

    /**
     * Tạo đơn hàng GHTK theo dữ liệu input từ client (Postman)
     * orderMap: là object "order" từ JSON
     * productsList: là danh sách products từ JSON
     */
    // ===================== 1️ ĐĂNG ĐƠN GHTK (Body build thủ công từ Map) =====================
    public Map<String, Object> registerOrder(Map<String, Object> orderMap, List<Map<String, Object>> productsList) {
        String url = BASE_URL + "shipment/order/?ver=1.5";
        Map<String, Object> body = new HashMap<>();
        body.put("order", orderMap);
        body.put("products", productsList);

        HttpEntity<Map<String, Object>> request = new HttpEntity<>(body, defaultHeaders());
        ResponseEntity<Map> response = restTemplate.postForEntity(url, request, Map.class);
        Map<String, Object> result = response.getBody();

        if (response.getStatusCode().is2xxSuccessful() && result != null && Boolean.TRUE.equals(result.get("success"))) {
            return (Map<String, Object>) result.get("order");
        } else {
            throw new RuntimeException("Không thể tạo đơn GHTK: " + (result != null ? result.get("message") : "Lỗi không xác định"));
        }
    }

    // ===================== 2️ ĐĂNG ĐƠN GHTK TỪ ENTITY (chuẩn GHTK) =====================
    public Map<String, Object> registerOrderWithProducts(ShipmentOrder shipmentOrder) {
        String url = BASE_URL + "shipment/order/?ver=1.5";

        Order order = shipmentOrder.getOrder();
        if (order == null) {
            throw new RuntimeException("ShipmentOrder chưa liên kết với Order");
        }
        UserAddress address = order != null ? order.getShippingAddress() : null;
        if (address == null || address.getAddress() == null) {
            throw new RuntimeException("Order chưa có địa chỉ người nhận hợp lệ");
        }

        // ------------------ Tạo Map order ------------------
//        Map<String, Object> orderMap = new HashMap<>();
//        orderMap.put("id", shipmentOrder.getId().toString());
//        orderMap.put("pick_name", "Kho tổng SmartMall");
//        orderMap.put("pick_address", "590 Cách Mạng Tháng 8, Phường 11, Quận 3");
//        orderMap.put("pick_province", "TP. Hồ Chí Minh");
//        orderMap.put("pick_district", "Quận 3");
//        orderMap.put("pick_ward", "Phường 11");
//        orderMap.put("pick_tel", "0911222333");
        // ------------------ Tạo Map order ------------------
        Map<String, Object> orderMap = new HashMap<>();
        orderMap.put("id", shipmentOrder.getId().toString());

        // Lấy thông tin kho từ ShipmentOrder hoặc cấu hình (ví dụ cột warehouseName)
        orderMap.put("pick_name", shipmentOrder.getWarehouse().getName() != null ?
                shipmentOrder.getWarehouse().getName()  : "Kho tổng SmartMall");
        orderMap.put("pick_address", shipmentOrder.getWarehouse().getAddress() != null ?
                shipmentOrder.getWarehouse().getAddress() : "590 Cách Mạng Tháng 8, Phường 11, Quận 3");
        orderMap.put("pick_province", shipmentOrder.getWarehouse().getProvince() != null ?
                shipmentOrder.getWarehouse().getProvince() : "TP. Hồ Chí Minh");
        orderMap.put("pick_district", shipmentOrder.getWarehouse().getDistrict() != null ?
                shipmentOrder.getWarehouse().getDistrict() : "Quận 3");
        orderMap.put("pick_ward", shipmentOrder.getWarehouse().getWard() != null ?
                shipmentOrder.getWarehouse().getWard()  : "Phường 11");
        orderMap.put("pick_tel", shipmentOrder.getWarehouse().getPhone() != null ?
                shipmentOrder.getWarehouse().getPhone() : "0911222333");

        // ------------------ Thông tin người nhận ------------------
        if (address != null && address.getAddress() != null) {
            Address addr = address.getAddress();
            orderMap.put("name", address.getUser().getProfile().getFullName());
            orderMap.put("tel", address.getPhoneNumber());
            orderMap.put("address", addr.getStreet());
            orderMap.put("province", addr.getCity());
            orderMap.put("district", addr.getDistrict());
            orderMap.put("ward", addr.getCommune());
            orderMap.put("hamlet", "Khác"); // bắt buộc nếu không có street
        } else {
            orderMap.put("name", "Nguyễn Văn A");
            orderMap.put("tel", "0911222333");
            orderMap.put("address", "123 Nguyễn Chí Thanh");
            orderMap.put("province", "TP. Hồ Chí Minh");
            orderMap.put("district", "Quận 1");
            orderMap.put("ward", "Phường Bến Nghé");
            orderMap.put("hamlet", "Khác");
        }
        double shippingFee =  order.getShippingFee() != null ? order.getShippingFee() : 0.0;
        orderMap.put("is_freeship", shippingFee == 0.0 ? "1" : "0");
        orderMap.put("pick_money", shipmentOrder.getCodAmount() != null ? shipmentOrder.getCodAmount() : BigDecimal.ZERO);
        orderMap.put("value",  order.getTotalAmount() != null ? BigDecimal.valueOf(order.getTotalAmount()) : BigDecimal.ZERO);
        orderMap.put("note", "Đơn test tạo từ SmartMall");
        orderMap.put("transport", "fly");
        orderMap.put("pick_option", "cod");
        orderMap.put("weight_option", "kg"); // đảm bảo GHTK hiểu kg

        // ------------------ Clone sản phẩm sang list mutable và check weight ------------------
        List<Map<String, Object>> productsList = new ArrayList<>();
        double totalWeight = 0.0;

        for (OrderItem item : order.getItems()) {
            ProductVariant variant = item.getVariant();
            if (variant == null) {
                throw new RuntimeException("OrderItem " + item.getId() + " chưa liên kết ProductVariant");
            }

            double weight = variant.getWeight() != null ? variant.getWeight() : 0.1;
            if (weight > 20) {
                throw new RuntimeException("Sản phẩm " + variant.getProduct().getName() + " vượt quá khối lượng tối đa 20kg");
            }

            totalWeight += weight * (item.getQuantity() != null ? item.getQuantity() : 1);

            Map<String, Object> p = new HashMap<>();
            p.put("name", variant.getProduct() != null && variant.getProduct().getName() != null
                    ? variant.getProduct().getName() : "Unnamed Product");
            p.put("weight", weight); // kg
            p.put("quantity", item.getQuantity() != null ? item.getQuantity() : 1);
            p.put("product_code", variant.getSku() != null ? variant.getSku() : "");
            productsList.add(p);
        }

        if (totalWeight >= 20) {
            throw new RuntimeException("Tổng khối lượng đơn hàng >= 20kg, không thể gửi GHTK");
        }

        orderMap.put("total_weight", totalWeight);

        // ------------------ Build body ------------------
        Map<String, Object> body = new HashMap<>();
        body.put("order", orderMap);
        body.put("products", productsList);

        System.out.println("===== BODY GỬI GHTK =====");
        System.out.println(body);
        System.out.println("=========================");

        // ------------------ Gửi request ------------------
        HttpEntity<Map<String, Object>> request = new HttpEntity<>(body, defaultHeaders());
        ResponseEntity<Map> response = restTemplate.postForEntity(url, request, Map.class);
        Map<String, Object> result = response.getBody();

        if (response.getStatusCode().is2xxSuccessful() && result != null && Boolean.TRUE.equals(result.get("success"))) {
            Map<String, Object> data = (Map<String, Object>) result.get("order");

            shipmentOrder.setTrackingCode((String) data.get("label"));
            shipmentOrder.setStatus(ShipmentStatus.REGISTERED);
            shipmentOrderRepository.save(shipmentOrder);

            log(shipmentOrder, "Tạo đơn GHTK thành công: " + data.get("label"));
            return data;
        } else {
            String msg = result != null ? result.get("message").toString() : "Không rõ lỗi từ GHTK";
            log(shipmentOrder, "Tạo đơn GHTK thất bại: " + msg);
            throw new RuntimeException("Không thể tạo đơn GHTK: " + msg);
        }
    }
    public ShipmentStatus fetchAndUpdateOrderStatus(String trackingCode) {
        String url = BASE_URL + "shipment/v2/" + trackingCode;
        HttpEntity<Void> request = new HttpEntity<>(defaultHeaders());

        ResponseEntity<Map> response = restTemplate.exchange(url, HttpMethod.GET, request, Map.class);
        Map<String, Object> result = response.getBody();

        if (response.getStatusCode().is2xxSuccessful() && result != null && Boolean.TRUE.equals(result.get("success"))) {
            Map<String, Object> orderData = (Map<String, Object>) result.get("order");

            // Lấy status từ GHTK và chuyển sang String
            Object statusObj = orderData.get("status");
            String ghtkStatus = statusObj != null ? statusObj.toString() : "1"; // mặc định "1" = PENDING

            // Map sang enum của app
            ShipmentStatus status = mapGhtkStatusToEnum(ghtkStatus);

            ShipmentOrder order = shipmentOrderRepository.findByTrackingCode(trackingCode)
                    .orElseThrow(() -> new RuntimeException("Không tìm thấy đơn hàng với mã: " + trackingCode));
            order.setStatus(status);
            shipmentOrderRepository.save(order);

            log(order, "Cập nhật trạng thái đơn hàng thành: " + status + " (GHTK status: " + ghtkStatus + ")");
            return status;
        } else {
            throw new RuntimeException("Không thể lấy trạng thái từ GHTK: " + (result != null ? result.get("message") : ""));
        }
    }

    /**
     * Map status GHTK (string) sang enum ShipmentStatus của app
     */
    private ShipmentStatus mapGhtkStatusToEnum(String ghtkStatus) {
        switch (ghtkStatus) {
            case "1": return ShipmentStatus.PENDING;      // Chưa tiếp nhận
            case "2": return ShipmentStatus.PICKING_UP;   // Đang lấy hàng
            case "3": return ShipmentStatus.IN_TRANSIT;   // Đang trung chuyển
            case "4": return ShipmentStatus.DELIVERED;    // Đã giao
            case "5": return ShipmentStatus.CANCELLED;    // Hủy
            default: return ShipmentStatus.PENDING;
        }
    }


    /** ------------------------- 3️ HỦY ĐƠN HÀNG --------------------------- **/
    public void cancelOrder(String trackingCode) {
        String url = BASE_URL + "shipment/cancel/" + trackingCode;
        HttpEntity<Void> request = new HttpEntity<>(defaultHeaders());

        ResponseEntity<Map> response = restTemplate.exchange(url, HttpMethod.POST, request, Map.class);
        Map<String, Object> result = response.getBody();

        if (response.getStatusCode().is2xxSuccessful() && result != null && Boolean.TRUE.equals(result.get("success"))) {
            ShipmentOrder order = shipmentOrderRepository.findByTrackingCode(trackingCode)
                    .orElseThrow(() -> new RuntimeException("Không tìm thấy đơn hàng: " + trackingCode));
            order.setStatus(ShipmentStatus.CANCELLED);
            shipmentOrderRepository.save(order);
            log(order, "Hủy đơn hàng thành công");
        } else {
            log(null, "Hủy đơn hàng GHTK thất bại: " + (result != null ? result.get("message") : "Không rõ lỗi"));
            throw new RuntimeException("Không thể hủy đơn GHTK");
        }
    }

    /** ------------------------- 4️ IN NHÃN VẬN ĐƠN ----------------------- **/
    public byte[] printLabel(String trackingCode) {
        String url = BASE_URL + "label/" + trackingCode + "?original=portrait&page_size=A6";

        HttpHeaders headers = new HttpHeaders();
        headers.set("Token", TOKEN);
        headers.set("X-Client-Source", PARTNER_CODE);

        HttpEntity<Void> request = new HttpEntity<>(headers);

        ResponseEntity<byte[]> response = restTemplate.exchange(
                url,
                HttpMethod.GET,
                request,
                byte[].class
        );

        if (response.getStatusCode().is2xxSuccessful() && response.getBody() != null) {
            log(null, "In nhãn vận đơn thành công cho mã: " + trackingCode);
            return response.getBody(); // Trả về PDF
        } else {
            throw new RuntimeException("Không thể in nhãn vận đơn, status: " + response.getStatusCode());
        }
    }

    /** ------------------------- 5️ TÍNH PHÍ VẬN CHUYỂN ------------------- **/
    public BigDecimal calculateShippingFee(String pickAddress, String deliverAddress, BigDecimal weight) {
        String url = BASE_URL + "shipment/fee";
        Map<String, Object> params = new HashMap<>();
        params.put("pick_address", pickAddress);
        params.put("address", deliverAddress);
        params.put("weight", weight);

        HttpEntity<Map<String, Object>> request = new HttpEntity<>(params, defaultHeaders());
        ResponseEntity<Map> response = restTemplate.postForEntity(url, request, Map.class);

        if (response.getStatusCode().is2xxSuccessful() && response.getBody() != null) {
            Map<String, Object> body = response.getBody();
            if (Boolean.TRUE.equals(body.get("success"))) {
                Map<String, Object> feeData = (Map<String, Object>) body.get("fee");
                BigDecimal fee = new BigDecimal(feeData.get("fee").toString());
                log(null, "Tính phí vận chuyển thành công: " + fee);
                return fee;
            }
        }
        throw new RuntimeException("Không thể tính phí GHTK");
    }

    /** ------------------------- 6️ HỖ TRỢ: GHI LOG ------------------------ **/
    private void log(ShipmentOrder order, String message) {
        try {
            ShipmentLog log = new ShipmentLog();
            log.setShipmentOrder(order);
            log.setMessage(message);
            log.setTimestamp(LocalDateTime.now());
            shipmentLogRepository.save(log);
        } catch (Exception e) {
            System.err.println("Lỗi ghi log GHTK: " + e.getMessage());
        }
    }
}

File: smart-mall-nextapp/src/services/GhtkService.ts
import apiClient from '../lib/apiClient';

export enum ShipmentStatus {
  PENDING = 'PENDING',
  PICKING_UP = 'PICKING_UP',
  IN_TRANSIT = 'IN_TRANSIT',
  DELIVERED = 'DELIVERED',
  RETURNING = 'RETURNING',
  RETURNED = 'RETURNED',
  CANCELLED = 'CANCELLED'
}

export interface GhtkOrderRequest {
  id: string;
  pick_name: string;
  pick_address: string;
  pick_province: string;
  pick_district: string;
  pick_ward: string;
  pick_tel: string;
  name: string; // Tên người nhận
  tel: string; // SĐT người nhận
  address: string; // Địa chỉ người nhận
  province: string;
  district: string;
  ward: string;
  hamlet?: string;
  is_freeship: string; // "1" hoặc "0"
  pick_money: number; // COD amount
  value: number; // Giá trị đơn hàng
  note?: string;
  transport?: string; // "road" | "fly"
  pick_option?: string; // "cod" | "post"
  weight_option?: string; // "kg" | "gram"
  total_weight?: number;
}

export interface GhtkProduct {
  name: string;
  weight: number; // kg
  quantity: number;
  product_code?: string;
}

export interface GhtkOrderResponse {
  label: string; // tracking code
  partner_id: string;
  status: string;
  status_text: string;
  created?: string;
  modified?: string;
  message?: string;
  pick_money: number;
  total_fee: number;
}

export interface GhtkOrderStatus {
  label: string;
  status: string;
  status_text: string;
  created: string;
  modified: string;
  message: string;
  pick_money: number;
  total_fee: number;
  description: string;
}

export interface GhtkFeeRequest {
  pick_address: string;
  address: string; // Địa chỉ giao hàng
  weight: number;
  value?: number;
  transport?: string;
}

export interface GhtkFeeResponse {
  fee: number;
  insurance_fee: number;
  include_vat: number;
  cost_id: string;
  delivery_type: string;
  a: number;
  dt: number;
  extFees: any[];
}

export class GhtkService {
  private static readonly BASE_URL = '/ghtk'; // Proxy qua backend của bạn
  
  // 1️⃣ Tạo đơn hàng GHTK với dữ liệu thủ công
  static async registerOrder(
    orderData: GhtkOrderRequest,
    products: GhtkProduct[]
  ): Promise<GhtkOrderResponse> {
    const response = await apiClient.post<{
      success: boolean;
      order: GhtkOrderResponse;
      message?: string;
    }>(`${this.BASE_URL}/register-order`, {
      order: orderData,
      products: products
    });

    if (response.data.success && response.data.order) {
      return response.data.order;
    } else {
      throw new Error(`Không thể tạo đơn GHTK: ${response.data.message || 'Lỗi không xác định'}`);
    }
  }

  // 2️⃣ Tạo đơn hàng GHTK từ ShipmentOrder entity
  static async registerOrderFromShipment(shipmentOrderId: string): Promise<GhtkOrderResponse> {
    try {
      const response = await apiClient.post<any>(`${this.BASE_URL}/register/${shipmentOrderId}`);
      
      // Backend trả về trực tiếp object GHTK response
      const data = response.data;
      
      // Kiểm tra nếu có label (thành công)
      if (data && data.label) {
        return {
          label: data.label,
          partner_id: data.partner_id || '',
          status: data.status_id?.toString() || '',
          status_text: data.status_text || '',
          pick_money: data.pick_money || 0,
          total_fee: data.fee || 0,
          message: 'Đăng ký GHTK thành công'
        };
      }
      
      // Nếu có cấu trúc wrapped
      if (data.success && data.order) {
        return data.order;
      }
      
      throw new Error(data.message || 'Không thể tạo đơn GHTK');
    } catch (error: any) {
      console.error('GHTK registration error:', error);
      throw new Error(
        error.response?.data?.message || 
        error.message || 
        'Không thể đăng ký GHTK'
      );
    }
  }

  // 3️⃣ Lấy và cập nhật trạng thái đơn hàng
  static async fetchAndUpdateOrderStatus(trackingCode: string): Promise<ShipmentStatus> {
    const response = await apiClient.get<{
      success: boolean;
      order: GhtkOrderStatus;
      status: ShipmentStatus;
      message?: string;
    }>(`${this.BASE_URL}/status/${trackingCode}`);

    if (response.data.success) {
      return response.data.status;
    } else {
      throw new Error(`Không thể lấy trạng thái từ GHTK: ${response.data.message || ''}`);
    }
  }

  // 4️⃣ Hủy đơn hàng
  static async cancelOrder(trackingCode: string): Promise<void> {
    const response = await apiClient.post<{
      success: boolean;
      message?: string;
    }>(`${this.BASE_URL}/cancel/${trackingCode}`);

    if (!response.data.success) {
      throw new Error(`Không thể hủy đơn GHTK: ${response.data.message || 'Lỗi không xác định'}`);
    }
  }

  // 5️⃣ In nhãn vận đơn
  static async printLabel(trackingCode: string): Promise<Blob> {
    const response = await apiClient.get(`${this.BASE_URL}/print/${trackingCode}`, {
      responseType: 'blob'
    });
    return response.data;
  }

  // 6️⃣ Tính phí vận chuyển
  static async calculateShippingFee(feeRequest: GhtkFeeRequest): Promise<GhtkFeeResponse> {
    const response = await apiClient.post<{
      success: boolean;
      fee: GhtkFeeResponse;
      message?: string;
    }>(`${this.BASE_URL}/calculate-fee`, feeRequest);

    if (response.data.success && response.data.fee) {
      return response.data.fee;
    } else {
      throw new Error(`Không thể tính phí GHTK: ${response.data.message || 'Lỗi không xác định'}`);
    }
  }

  // 7️⃣ Lấy thông tin chi tiết đơn hàng
  static async getOrderDetails(trackingCode: string): Promise<GhtkOrderStatus> {
    const response = await apiClient.get<{
      success: boolean;
      order: GhtkOrderStatus;
      message?: string;
    }>(`${this.BASE_URL}/order-details/${trackingCode}`);

    if (response.data.success && response.data.order) {
      return response.data.order;
    } else {
      throw new Error(`Không thể lấy thông tin đơn hàng: ${response.data.message || 'Lỗi không xác định'}`);
    }
  }

  // 8️⃣ Lấy danh sách địa chỉ (tỉnh/thành)
  static async getProvinces(): Promise<Array<{ name: string; code: string }>> {
    const response = await apiClient.get<{
      success: boolean;
      data: Array<{ name: string; code: string }>;
    }>(`${this.BASE_URL}/provinces`);

    return response.data.data || [];
  }

  // 9️⃣ Lấy danh sách quận/huyện
  static async getDistricts(provinceCode: string): Promise<Array<{ name: string; code: string }>> {
    const response = await apiClient.get<{
      success: boolean;
      data: Array<{ name: string; code: string }>;
    }>(`${this.BASE_URL}/districts/${provinceCode}`);

    return response.data.data || [];
  }

  // 🔟 Lấy danh sách phường/xã
  static async getWards(districtCode: string): Promise<Array<{ name: string; code: string }>> {
    const response = await apiClient.get<{
      success: boolean;
      data: Array<{ name: string; code: string }>;
    }>(`${this.BASE_URL}/wards/${districtCode}`);

    return response.data.data || [];
  }

  // 📊 Thống kê đơn hàng GHTK
  static async getOrderStatistics(
    startDate?: string,
    endDate?: string
  ): Promise<{
    totalOrders: number;
    deliveredOrders: number;
    pendingOrders: number;
    cancelledOrders: number;
    totalFee: number;
    totalCod: number;
  }> {
    const params = new URLSearchParams();
    if (startDate) params.append('startDate', startDate);
    if (endDate) params.append('endDate', endDate);

    const response = await apiClient.get<{
      success: boolean;
      statistics: {
        totalOrders: number;
        deliveredOrders: number;
        pendingOrders: number;
        cancelledOrders: number;
        totalFee: number;
        totalCod: number;
      };
    }>(`${this.BASE_URL}/statistics?${params.toString()}`);

    return response.data.statistics;
  }

  // 🔄 Đồng bộ trạng thái tất cả đơn hàng
  static async syncAllOrderStatus(): Promise<{
    updated: number;
    failed: number;
  }> {
    const response = await apiClient.post<{
      success: boolean;
      result: {
        updated: number;
        failed: number;
      };
    }>(`${this.BASE_URL}/sync-all-status`);

    return response.data.result;
  }

  // 🏷️ Tạo nhãn hàng loạt
  static async printMultipleLabels(trackingCodes: string[]): Promise<Blob> {
    const response = await apiClient.post(`${this.BASE_URL}/print-multiple-labels`, {
      trackingCodes
    }, {
      responseType: 'blob'
    });
    return response.data;
  }

  // ⚙️ Helper Methods

  // Map GHTK status sang enum của app
  static mapGhtkStatusToEnum(ghtkStatus: string): ShipmentStatus {
    switch (ghtkStatus) {
      case '1': return ShipmentStatus.PENDING;      // Chưa tiếp nhận
      case '2': return ShipmentStatus.PICKING_UP;   // Đang lấy hàng
      case '3': return ShipmentStatus.IN_TRANSIT;   // Đang trung chuyển
      case '4': return ShipmentStatus.DELIVERED;    // Đã giao
      case '5': return ShipmentStatus.CANCELLED;    // Hủy
      default: return ShipmentStatus.PENDING;
    }
  }

  // Validate trọng lượng đơn hàng
  static validateOrderWeight(products: GhtkProduct[]): {
    isValid: boolean;
    totalWeight: number;
    errors: string[];
  } {
    const errors: string[] = [];
    let totalWeight = 0;

    for (const product of products) {
      const productWeight = product.weight * product.quantity;
      totalWeight += productWeight;

      if (product.weight > 20) {
        errors.push(`Sản phẩm "${product.name}" vượt quá khối lượng tối đa 20kg`);
      }
    }

    if (totalWeight >= 20) {
      errors.push('Tổng khối lượng đơn hàng >= 20kg, không thể gửi GHTK');
    }

    return {
      isValid: errors.length === 0,
      totalWeight,
      errors
    };
  }

  // Format trạng thái hiển thị
  static formatStatus(status: ShipmentStatus): string {
    const statusMap: Record<ShipmentStatus, string> = {
      [ShipmentStatus.PENDING]: 'Chờ xử lý',
      [ShipmentStatus.PICKING_UP]: 'Đang lấy hàng',
      [ShipmentStatus.IN_TRANSIT]: 'Đang vận chuyển',
      [ShipmentStatus.DELIVERED]: 'Đã giao hàng',
      [ShipmentStatus.RETURNING]: 'Đang hoàn trả',
      [ShipmentStatus.RETURNED]: 'Đã hoàn trả',
      [ShipmentStatus.CANCELLED]: 'Đã hủy',
    };
    return statusMap[status] || status;
  }

  // Format tiền tệ
  static formatCurrency(amount: number): string {
    return new Intl.NumberFormat('vi-VN', {
      style: 'currency',
      currency: 'VND'
    }).format(amount);
  }

  // Tạo địa chỉ đầy đủ
  static buildFullAddress(
    street: string,
    ward: string,
    district: string,
    province: string
  ): string {
    return `${street}, ${ward}, ${district}, ${province}`;
  }

  // Validate địa chỉ
  static validateAddress(orderData: GhtkOrderRequest): {
    isValid: boolean;
    errors: string[];
  } {
    const errors: string[] = [];

    if (!orderData.pick_address) errors.push('Địa chỉ lấy hàng là bắt buộc');
    if (!orderData.pick_province) errors.push('Tỉnh/thành lấy hàng là bắt buộc');
    if (!orderData.pick_district) errors.push('Quận/huyện lấy hàng là bắt buộc');
    if (!orderData.pick_ward) errors.push('Phường/xã lấy hàng là bắt buộc');
    if (!orderData.pick_tel) errors.push('SĐT lấy hàng là bắt buộc');

    if (!orderData.name) errors.push('Tên người nhận là bắt buộc');
    if (!orderData.tel) errors.push('SĐT người nhận là bắt buộc');
    if (!orderData.address) errors.push('Địa chỉ giao hàng là bắt buộc');
    if (!orderData.province) errors.push('Tỉnh/thành giao hàng là bắt buộc');
    if (!orderData.district) errors.push('Quận/huyện giao hàng là bắt buộc');
    if (!orderData.ward) errors.push('Phường/xã giao hàng là bắt buộc');

    return {
      isValid: errors.length === 0,
      errors
    };
  }
}

export default GhtkService;